### DISCUSSION TITLE
`StateMachine<_, Data>` states don't have access to the type of `Data`

### DISCUSSION THREAD

**Discussion by danneu** [excaliburjs/Excalibur#3277] [General]
(Btw, I love this game engine. Thanks for all the work.)

Reference: [src/engine/Util/StateMachine.ts](https://github.com/excaliburjs/Excalibur/blob/80205605ecb4ac583aa404a72c6b510af6d600a0/src/engine/Util/StateMachine.ts#L10)

consider this state machine:

```typescript
import { StateMachine, StateMachineDescription } from "excalibur";

type Data = { counter: number }

const data: Data = { counter: 0 };

const description: StateMachineDescription = {
  start: "Idle",
  states: {
    Idle: {
      transitions: [],
      onEnter: ({ data }) => { // data is `any`
        console.log("data is", data);
      },
    },
  },
};

const stateMachine: StateMachine<'Idle', Data> = StateMachine.create(description, { counter: 0 });
```

Since  neither `StateMachineDescription` nor `State` have any type parameters, they can't know the `Data` parameter of the StateMachine that they're used in. 

Instead I expected the types to be defined `StateMachineDescription<States, Data>` and `State<Data>` which would have to match `StateMachine<States, Data>` so that you can write type-safe states.

Aside, shouldn't there be a `stateMachine.start(data)` method to kick off the machine starting at `startingState`?

**Comment by eonarheim** [comment]
@danneu Thanks for the kind words, I tweaked the types on the main line branch to type the data as you expect. https://github.com/excaliburjs/Excalibur/commit/dccb762a4e136a55706f3e36b2522fcf3f328a4e

For a `start()` method I'll need to think a little longer, but makes sense. Currently the machine is "implicitly" started when created.

**Reply by eonarheim** [reply]
@danneu this should be live now in the latest v0.30.x!
